corehq/apps/app_manager/static/app_manager/js/report-module.js

```diff
@@ -1,17 +1,108 @@
 
 var ReportModule = (function () {
 
-    function ReportConfig(report_id, display, availableReportIds, language) {
+    function Config(dict) {
+        var self = this;
+
+        dict = dict || {};
+
+        self.keyValuePairs = ko.observableArray();
+        for (var key in dict) {
+            self.keyValuePairs.push([ko.observable(key), ko.observable(dict[key])]);
+        }
+
+        self.addConfig = function() {
+            self.keyValuePairs.push([ko.observable(''), ko.observable('')]);
+        };
+    }
+
+    function GraphConfig(report_id, reportId, availableReportIds, reportCharts, graph_configs) {
+        var self = this;
+
+        graph_configs = graph_configs || {};
+
+        this.graphConfigs = {};
+        for (var i = 0; i < availableReportIds.length; i++) {
+            var currentReportId = availableReportIds[i];
+            self.graphConfigs[currentReportId] = {};
+            for (var j = 0; j < reportCharts[currentReportId].length; j++) {
+                var currentChart = reportCharts[currentReportId][j];
+                var graph_config = graph_configs[currentChart.chart_id] || {};
+                var series_configs = {};
+                var chart_series = [];
+                for(var k = 0; k < currentChart.y_axis_columns.length; k++) {
+                    var series = currentChart.y_axis_columns[k];
+                    chart_series.push(series);
+                    series_configs[series] = new Config(
+                        currentReportId == report_id ? (graph_config.series_configs || {})[series] || {} : {}
+                    );
+                }
+
+                self.graphConfigs[currentReportId][currentChart.chart_id] = {
+                    graph_type: ko.observable(currentReportId == report_id ? graph_config.graph_type || 'bar' : 'bar'),
+                    series_configs: series_configs,
+                    chart_series: chart_series,
+                    config: new Config(
+                        currentReportId == report_id ? graph_config.config || {} : {}
+                    )
+                };
+            }
+        }
+
+        this.currentGraphConfigs = ko.computed(function() {
+            return self.graphConfigs[reportId()];
+        });
+
+        this.currentCharts = ko.computed(function() {
+            return reportCharts[reportId()];
+        });
+
+        this.getCurrentGraphConfig = function(chart_id) {
+            return self.currentGraphConfigs()[chart_id] || {config: {keyValuePairs: []}};
+        };
+
+        this.toJSON = function () {
+            function configToDict(config) {
```

Comment:
for fun, in underscore this is a one-liner: `_.object(_.partition(config.keyValuePairs(), function(x, i) { return i % 2 == 0; }))`

maybe also with a `map` to call each key & value, and granted I'm not 100% sure that underscore can iterate through an observableArray